components/ILIAS/Mail/classes/class.ilMailCronTerminatedMails.php

```diff
@@ -0,0 +1,177 @@
+            $this->mail = new ilMail($this->user->getId());
+            $this->umail = new ilFormatMail($this->user->getId());
+
+            $this->lng->loadLanguageModule('mail');
+            $this->init_done = true;
+        }
+    }
+
+    public function getId(): string
+    {
+        return 'mail_terminated_mails';
+    }
+
+    public function getTitle(): string
+    {
+        $this->init();
+
+        return $this->lng->txt('mail_cron_terminated_mails');
+    }
+
+    public function getDescription(): string
+    {
+        $this->init();
+
+        return $this->lng->txt('mail_cron_terminated_mails_desc');
+    }
+
+    public function hasAutoActivation(): bool
+    {
+        return true;
+    }
+
+    public function hasFlexibleSchedule(): bool
+    {
+        return true;
+    }
+
+    public function getDefaultScheduleType(): JobScheduleType
+    {
+        return JobScheduleType::DAILY;
+    }
+
+    public function getDefaultScheduleValue(): ?int
+    {
+        return 1;
+    }
+
+    public function run(): JobResult
+    {
+        $this->init();
+
+        $job_result = new JobResult();
+        $job_result->setStatus(JobResult::STATUS_OK);
+
+        ilLoggerFactory::getLogger('mail')->info('Start sending terminated mails from all users.');
+
+        $mails = $this->getOutboxMails();
+        $sent_mail_ids = [];
+        foreach ($mails as $mail) {
+            try {
+                $mailer = $this->umail
+                    ->withContextId(ilContext::CONTEXT_CRON);
+
+                $mailer->setSaveInSentbox(true);
+
+                $mailer->autoresponder()->enableAutoresponder();
+                $errors = $mailer->enqueue(
+                    $mail['rcp_to'],
+                    $mail['rcp_cc'],
+                    $mail['rcp_bcc'],
+                    $mail['m_subject'],
+                    $mail['m_message'],
+                    unserialize($mail['attachments'], ['allowed_classes' => false]),
+                    (bool) ($mail['use_placeholders'] ?? false)
+                );
+
+                if (empty($errors)) {
+                    $sent_mail_ids[] = (int) ($mail['mail_id'] ?? 0);
+                }
+            } catch (\Exception $e) {
+                $job_result->setStatus(JobResult::STATUS_FAIL);
+                ilLoggerFactory::getLogger('mail')->error(
+                    'Error sending terminated mail with id ' . ($mail['mail_id'] ?? 'unknown') . ': ' . $e->getMessage()
+                );
+                $job_result->setMessage($e->getMessage());
+                return $job_result;
+            }
+        }
+        $this->mail->deleteMails($sent_mail_ids);
+        ilLoggerFactory::getLogger('mail')->info(
+            'Sent ' . count($sent_mail_ids) . ' terminated mails and removed them from outbox.'
+        );
+        $job_result->setMessage('Processed ' . count($sent_mail_ids) . ' mails.');
+        return $job_result;
+    }
+
+    public function getOutboxMails(): Generator
```

Comment:
It would be good if we could describe the shape of the array and use this type in the Genertor.

```php
...
/**
 * @phpstan-type MailWithFolderDatabaseRecordType array{
 *     obj_id: int,
 * }
 */
class ilMail
{
...
```
AI could really help to define such record types.

You could then use `\Generator<MailWithFolderDatabaseRecordType>` and import  this type into consuming classes:

```php
...
/**
 * @phpstan-import-type MailWithFolderDatabaseRecordType from \ilMail
 */
class MailConsumer
{
...
```